### PR TITLE
Use nativeloader instead of system loader to load JNI library for soloader compatibility.

### PR DESCRIPTION
**Pull request by xcheng16**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29350 [Pytorch] system loader to native loader.**

Differential Revision: [D18352773](https://our.internmc.facebook.com/intern/diff/D18352773/)

### CODE REVIEW

File: android/pytorch_android/CMakeLists.txt [hunk 1]
```diff
@@ -62,10 +62,8 @@ if (ANDROID_ABI)
     # Link most things statically on Android.
   target_link_libraries(pytorch
       fbjni
-      -Wl,--gc-sections
-      -Wl,--whole-archive
+      -Wl,-dead_strip
       libtorch
-      -Wl,--no-whole-archive
```
Comment (dreiss) [inline android/pytorch_android/CMakeLists.txt]: This is for the Android build.  I don't think it needs to change.

File: android/pytorch_android/CMakeLists.txt [hunk 2]
```diff
@@ -79,10 +77,8 @@ else()
   # Prefer dynamic linking on the host
   target_link_libraries(pytorch
       fbjni
-      -Wl,--gc-sections
-      -Wl,--whole-archive
+      -Wl,-dead_strip
       torch
-      -Wl,--no-whole-archive
```
Comment (dreiss) [inline android/pytorch_android/CMakeLists.txt]: Unfortunately, I suspect this breaks the Linux build, which links with gold.  Any way we can get both Mac and Linux builds to work?

Also, this is not related to the subject of the diff.  Please use separate diffs for unrelated changes.

File: android/pytorch_android/src/androidTest/java/org/pytorch/PytorchHostTests.java [hunk 3]
```diff
@@ -1,28 +1,22 @@
 package org.pytorch;
 
-import com.facebook.soloader.nativeloader.NativeLoader;
```
Comment (dreiss) [inline android/pytorch_android/src/androidTest/java/org/pytorch/PytorchHostTests.java]: It looks like this file was reformatted.  Can you do that as a separate diff first to make this change more self-contained?

File: android/pytorch_android/src/androidTest/java/org/pytorch/PytorchHostTests.java [hunk 4]
```diff
@@ -1,28 +1,22 @@
...
+import org.junit.BeforeClass;
 
 public class PytorchHostTests extends PytorchTestBase {
   @BeforeClass
-  public static void setUpClass() {
-    if (!NativeLoader.isInitialized()) {
-      NativeLoader.init(new SystemDelegate());
-    }
-  }
+  public static void setUpClass() {}
```
Comment (dreiss) [inline android/pytorch_android/src/androidTest/java/org/pytorch/PytorchHostTests.java]: No longer needed.

File: android/pytorch_android/src/androidTest/java/org/pytorch/PytorchTestBase.java [hunk 5]
```diff
@@ -1,30 +1,26 @@
...
+import org.junit.Test;
+
 public abstract class PytorchTestBase {
   private static final String TEST_MODULE_ASSET_NAME = "test.pt";
 
   @Before
-  public void setUp() {
-    System.loadLibrary("pytorch");
-  }
+  public void setUp() {}
```
Comment (dreiss) [inline android/pytorch_android/src/androidTest/java/org/pytorch/PytorchTestBase.java]: No longer needed.

File: android/pytorch_android/src/androidTest/java/org/pytorch/PytorchTestBase.java [hunk 6]
```diff
@@ -61,17 +57,17 @@ public void testEqFloat() throws IOException {
...
+          Double.MAX_VALUE,
+          -Double.MIN_VALUE,
+          Double.MIN_VALUE,
+          -Math.exp(1.d),
+          -Math.sqrt(2.d),
+          -3.1415f,
+          3.1415f,
+          -1,
+          0,
+          1,
```
Comment (dreiss) [inline android/pytorch_android/src/androidTest/java/org/pytorch/PytorchTestBase.java]: Reformat should be a separate diff.

File: android/pytorch_android/src/main/cpp/pytorch_jni.cpp [hunk 7]
```diff
@@ -418,7 +418,7 @@ class JIValue : public facebook::jni::JavaClass<JIValue> {
     } else if (JIValue::kTypeCodeLong == typeCode) {
       static const auto jMethodGetLong =
           JIValue::javaClassStatic()->getMethod<jlong()>("toLong");
-      return at::IValue{jMethodGetLong(jivalue)};
+      return at::IValue{(int64_t)jMethodGetLong(jivalue)};
```
Comment (dreiss) [inline android/pytorch_android/src/main/cpp/pytorch_jni.cpp]: This is part of the Mac PR, right?

**Review by dreiss — CHANGES_REQUESTED** [review]
Nit: The "T" in PyTorch is capitalized.

Title shouldn't have "[PyTorch]" in GitHub (ghimport should automatically add it internally).

Title should probably clearly state "Use NativeLoader instead of system loader to load JNI library", and should say why we want to do this (so it's compatible with SoLoader).

File: android/pytorch_android/src/main/cpp/pytorch_jni.cpp (hunk 7 above)
Comment (xcheng16) [reply android/pytorch_android/src/main/cpp/pytorch_jni.cpp]: yes, It is for mac, I will create separate pr for it.

File: android/pytorch_android/src/androidTest/java/org/pytorch/PytorchTestBase.java (hunk 6 above)
Comment (xcheng16) [reply android/pytorch_android/src/androidTest/java/org/pytorch/PytorchTestBase.java]: when update the diff, it auto-format the file, i can revert it.
